src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php

```diff
@@ -211,6 +236,262 @@ public function deleteAction(Request $request, int $productId): Response
         return $this->redirectToRoute('admin_products_v2_index');
     }
 
+    /**
+     * @param Request $request
+     * @param int $productId
+     *
+     * @return Response
+     */
+    public function previewAction(Request $request, int $productId): Response
+    {
+        try {
+            $productUrlProvider = $this->get('prestashop.adapter.shop.url.product_provider');
+            $url = $productUrlProvider->getUrl($productId, '{friendly-url}');
+        } catch (ProductException $e) {
+            $this->addFlash('error', $this->getErrorMessageForException($e, $this->getErrorMessages($e)));
+
+            return $this->redirectToRoute('admin_products_v2_index');
+        }
+
+        return $this->redirect($url);
+    }
+
+    /**
+     * @param Request $request
+     * @param int $productId
+     *
+     * @return Response
+     */
+    public function duplicateAction(Request $request, int $productId): Response
+    {
+        try {
+            $this->getCommandBus()->handle(new DuplicateProductCommand($productId));
+            $this->addFlash(
+                'success',
+                $this->trans('Successful duplication', 'Admin.Notifications.Success')
+            );
+        } catch (ProductException $e) {
+            $this->addFlash('error', $this->getErrorMessageForException($e, $this->getErrorMessages($e)));
+        }
+
+        return $this->redirectToRoute('admin_products_v2_index');
+    }
+
+    /**
+     * Toggles product status
+     *
+     * @AdminSecurity("is_granted('update', request.get('_legacy_controller'))", redirectRoute="admin_products_v2_index")
+     * @DemoRestricted(redirectRoute="admin_products_v2_index")
+     *
+     * @param int $productId
+     *
+     * @return RedirectResponse
+     */
+    public function toggleStatusAction(int $productId): RedirectResponse
```

Comment:
This method actually already exists in the legacy controller and already uses the queries and commands But you can copy it here since the legacy controller is bound to be removed

The main difference is that it's an ajax action which returns a json response for its success or failure, and it should be used via an ajax request in the grid as well, it relies on a special ToggleColumn which has been used in the monitoring controller that displays products already:
https://github.com/PrestaShop/PrestaShop/blob/563fceacb97c86c8b2e38d1f71ed96159e46bd9d/src/Core/Grid/Definition/Factory/Monitoring/AbstractProductGridDefinitionFactory.php#L104

I believe it relies on a prestashop component that you can activate with the `initComponents` method

Reply:
This should be an ajax controller